intermediate_source/FSDP_adavnced_tutorial.rst

```diff
@@ -0,0 +1,534 @@
+        
+        cleanup()
+
+
+
+2.5 Finally parsing the arguments and setting the main function
+
+.. code-block:: python
+
+    
+    if __name__ == '__main__':
+    # Training settings
+    parser = argparse.ArgumentParser(description='PyTorch T5 FSDP Example')
+    parser.add_argument('--batch-size', type=int, default=4, metavar='N',
+                        help='input batch size for training (default: 64)')
+    parser.add_argument('--test-batch-size', type=int, default=4, metavar='N',
+                        help='input batch size for testing (default: 1000)')
+    parser.add_argument('--epochs', type=int, default=1, metavar='N',
+                        help='number of epochs to train (default: 14)')
+    parser.add_argument('--lr', type=float, default=1.0, metavar='LR',
+                        help='learning rate (default: 1.0)')
+    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
+                        help='Learning rate step gamma (default: 0.7)')
+    parser.add_argument('--no-cuda', action='store_true', default=False,
+                        help='disables CUDA training')
+    parser.add_argument('--seed', type=int, default=1, metavar='S',
+                        help='random seed (default: 1)')
+    parser.add_argument('--track_memory', action='store_true', default=False,
+                        help='track the gpy memory')
+    parser.add_argument('--run_validation', action='store_true', default=False,
+                        help='running the validation')
+    parser.add_argument('--activation_checkpointing', action='store_true', default=False,
+                        help='Checkpoint activations')
+    parser.add_argument('--save-model', action='store_true', default=False,
+                        help='For Saving the current Model')
+    args = parser.parse_args()
+
+    torch.manual_seed(args.seed)
+    
+    fsdp_main(args)
+
+
+To run the the training using torchrun:
+
+.. code-block:: bash 
+
+    torchrun --nnodes 1 --nproc_per_node 4  T5_training.py
+
+.. _transformer_wrapping_policy:
+Transformer Wrapping Policy
+---------------------------
+As discussed in the `previous tutorial <https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html>`__, fsdp_auto_wrap_policy is one of the FSDP features that make it easy to automatically shard a given model and put the model, optimizer and gradient shards into distinct FSDP units.
+
+For some architectures such as Transformer encoder-decoders, some part of the model such as embedding table is being shared with both encoder and decoder.
+In this case, we need to place the embedding table in the outer FSDP unit so that it could be accessed from both encoder and decoder.  In addition, by registering the layer class for a transformer, the sharding plan can be made much more communication efficient.  In Pytorch 1.12, FSDP added this support and now we have a wrapping policy for transfomers.
```

Comment:
Add something like "As of PyTorch 1.12, shared parameters such as this shared embedding table must be placed in the same FSDP unit, so all parameters which point to this embedding table must be part of the same FSDP unit".

Reply:
@rohan-varma , I wonder if this is different from having shared parameters in the outer FSDP unit, that could be accessed by other parts of the model?